File: edge-device/RGB/demo/rgb.py
#!/usr/bin/env python3
# encoding: utf-8

import RPi.GPIO as gpio
import time

R, G, B = 12, 13, 19

gpio.setmode(gpio.BCM)

gpio.setup(R, gpio.OUT)
gpio.setup(G, gpio.OUT)
gpio.setup(B, gpio.OUT)

pwmR = gpio.PWM(R, 50)
pwmG = gpio.PWM(G, 50)
pwmB = gpio.PWM(B, 50)

pwmR.start(0)
pwmG.start(0)
pwmB.start(0)

try:
    t = 0.3
    while True:
        # 红色灯全亮，蓝灯，绿灯全暗（红色）
        # lights red
        pwmR.ChangeDutyCycle(100)
        pwmG.ChangeDutyCycle(0)
        pwmB.ChangeDutyCycle(0)
        time.sleep(t)

        # 绿色灯全亮，红灯，蓝灯全暗（绿色）
        # lights green
        pwmR.ChangeDutyCycle(0)
        pwmG.ChangeDutyCycle(100)
        pwmB.ChangeDutyCycle(0)
        time.sleep(t)

        # 蓝色灯全亮，红灯，绿灯全暗（蓝色）
        # lights blue
        pwmR.ChangeDutyCycle(0)
        pwmG.ChangeDutyCycle(0)
        pwmB.ChangeDutyCycle(100)
        time.sleep(t)

        # 红灯，绿灯全亮，蓝灯全暗（黄色）
        # lights red and green => yellow
        pwmR.ChangeDutyCycle(100)
        pwmG.ChangeDutyCycle(100)
        pwmB.ChangeDutyCycle(0)
        time.sleep(t)

        # 红灯，蓝灯全亮，绿灯全暗（洋红色）
        # lights red and blue => magenta
        pwmR.ChangeDutyCycle(100)
        pwmG.ChangeDutyCycle(0)
        pwmB.ChangeDutyCycle(100)
        time.sleep(t)

        # 绿灯，蓝灯全亮，红灯全暗（青色）
        # lights green and blue => aqua
        pwmR.ChangeDutyCycle(0)
        pwmG.ChangeDutyCycle(100)
        pwmB.ChangeDutyCycle(100)
        time.sleep(t)

        # 红灯，绿灯，蓝灯全亮（白色）
        # lights red, green and blue => white
        pwmR.ChangeDutyCycle(100)
        pwmG.ChangeDutyCycle(100)
        pwmB.ChangeDutyCycle(100)
        time.sleep(t)

except KeyboardInterrupt:
    pass

pwmR.stop()
pwmG.stop()
pwmB.stop()

gpio.cleanup()

File: edge-device/RGB/demo/rgb.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/stianeikeland/go-rpio"
)

const cycleLength = 16
const pmwClockFrequency = 1000 * cycleLength
const timeSpace = 3000 // in ms

var onlyOneSignalHandler = make(chan struct{})
var shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}

func setupSignalHandler() (stopCh <-chan struct{}) {
	close(onlyOneSignalHandler) // panics when called twice
	stop := make(chan struct{})
	c := make(chan os.Signal, 1)
	signal.Notify(c, shutdownSignals...)
	go func() {
		<-c
		rpio.StopPwm()
		rpio.Close()
		fmt.Println("exit")
		os.Exit(1)
	}()
	return stop
}

func stoper(stopCh <-chan struct{}) {
	go func() {
		<-stopCh
	}()
}

func main() {
	stoper(setupSignalHandler())

	err := rpio.Open()
	if err != nil {
		os.Exit(1)
	}
	defer func() {
		fmt.Println("exit")
		rpio.StopPwm()
		rpio.Close()
	}()

	// for red led
	pinR := rpio.Pin(12)
	pinR.Mode(rpio.Pwm)
	pinR.Freq(pmwClockFrequency)
	pinR.DutyCycle(0, cycleLength)

	// for green led
	pinG := rpio.Pin(13)
	pinG.Mode(rpio.Pwm)
	pinG.Freq(pmwClockFrequency)
	pinG.DutyCycle(0, cycleLength)

	// for blue led
	pinB := rpio.Pin(19)
	pinB.Mode(rpio.Pwm)
	pinB.Freq(pmwClockFrequency)
	pinB.DutyCycle(0, cycleLength)

	for {
		fmt.Println("#F00") // in fact, it lights green
		pinR.DutyCycle(16, cycleLength)
		pinG.DutyCycle(0, cycleLength)
		pinB.DutyCycle(0, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#0F0") // in fact, it lights none
		pinR.DutyCycle(0, cycleLength)
		pinG.DutyCycle(16, cycleLength)
		pinB.DutyCycle(0, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#00F") // in fact, it lights red and blue
		pinR.DutyCycle(0, cycleLength)
		pinG.DutyCycle(0, cycleLength)
		pinB.DutyCycle(16, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#FF0") // in fact, it lights green
		pinR.DutyCycle(16, cycleLength)
		pinG.DutyCycle(16, cycleLength)
		pinB.DutyCycle(0, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#F0F") // in fact, it lights white
		pinR.DutyCycle(16, cycleLength)
		pinG.DutyCycle(0, cycleLength)
		pinB.DutyCycle(16, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#0FF") // in fact, it lights red and blue
		pinR.DutyCycle(0, cycleLength)
		pinG.DutyCycle(16, cycleLength)
		pinB.DutyCycle(16, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)

		fmt.Println("#FFF") // in fact, it lights white
		pinR.DutyCycle(16, cycleLength)
		pinG.DutyCycle(16, cycleLength)
		pinB.DutyCycle(16, cycleLength)
		time.Sleep(time.Millisecond * timeSpace)
	}
}
